README.md
# React minimal pie chart [![Build Status][ci-img]][ci]

INFORMATION: This is a highly customized version of "React minimal pie chart". It might not meet your requirements anymore. If so, please use the original libary at: https://github.com/toomuchdesign/react-minimal-pie-chart

Lightweight React **SVG pie charts**, with **versatile options** and **CSS animation** included. 👏 &nbsp;[Demo][storybook]&nbsp;👏.

[ci-img]: https://travis-ci.org/toomuchdesign/react-minimal-pie-chart.svg
[ci]: https://travis-ci.org/toomuchdesign/react-minimal-pie-chart
[storybook]: http://www.andreacarraro.it/react-minimal-pie-chart

```js
import PieChart from 'react-minimal-pie-chart'

;<PieChart data={[{ value: 10, color: '#E38627' }, { value: 15, color: '#C13C37' }, { value: 20, color: '#6A2135' }]} />
```

## Installation

```console
npm install @coduct/react-minimal-pie-chart
```

## Why?

Because [Recharts](https://github.com/recharts/recharts) is awesome, but when you need just a simple pie/donought chart, few line of code are usually enough.

## Features

- No dependencies (except for [React's prop-types](https://github.com/reactjs/prop-types))
- Customizable CSS animations trough [stroke-dasharray + stroke-dashoffset strategy](https://css-tricks.com/svg-line-animation-works/)
- Configurable: Pie, Donut, Loading, Completion charts (see [Demo][storybook])

## Options

| Property              | Type     | Description                                                                                                    | Default    |
| --------------------- | -------- | -------------------------------------------------------------------------------------------------------------- | ---------- |
| **data** _(required)_ | _Array_  | The source data which each element is a segment.                                                               | -          |
| **cx**                | _Number_ | The x-coordinate of center. The value is the percentage of the component width                                 | 50         |
| **cy**                | _Number_ | The y-coordinate of center. The value is the percentage of the component height                                | 50         |
| **ratio**             | _Number_ | The ratio of rendered svg element                                                                              | 1          |
| **startAngle**        | _Number_ | The start angle of first sector                                                                                | 0          |
| **lengthAngle**       | _Number_ | The total angle taken by the chart _(can be negative to make the chart clockwise!)_                            | 360        |
| **totalValue**        | _Number_ | The total value represented by the full chart                                                                  | -          |
| **lineWidth**         | _Number_ | The width of the line representing each sector. The value is the percentage of chart radio*(100 === full pie)* | 100        |
| **radius**            | _Number_ | The radius of the pie. The value is the percentage of the component width \*                                   | 50         |
| **paddingAngle**      | _Number_ | The angle between two sectors                                                                                  | -          |
| **rounded**           | _Bool_   | Round line caps of each sector                                                                                 | false      |
| **style**             | _Object_ | The style object assigned to chart wrapper                                                                     | -          |
| **animate**           | _Bool_   | Animate sectors on component mount                                                                             | false      |
| **animationDuration** | _Number_ | Animation duration in ms                                                                                       | 500        |
| **animationEasing**   | _String_ | Animation CSS easing                                                                                           | "ease-out" |
| **reveal**            | _Number_ | Turn on CSS animation and reveal just a percentage of each segment                                             | -          |

Each **data** entry can also accept an **optional [`key` property](https://reactjs.org/docs/lists-and-keys.html)** just in case items' indexes weren't enough:

```js
{ value: 10, key: 1, color: '#E38627' }
```

## Browsers support

The main requirement of this library is an accurate rendering of [SVG Stroke properties](https://www.w3schools.com/graphics/svg_stroking.asp).

### Not supported

- IE ≤ 10

### Partially supported

- IE 11

## Misc

### How svg arc paths work?

http://users.ecs.soton.ac.uk/rfp07r/interactive-svg-examples/

## Todo's

- Find a better `paddingAngle` implementation
- Make a device/browser compatibility table
- Define a className targeting each segment path for custom CSS animations
- Background segment
- Add `babel-eslint`?
- Add UMD export?

## Contributors

Thanks to you all ([emoji key](https://github.com/kentcdodds/all-contributors#emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->

| [<img src="https://avatars3.githubusercontent.com/u/4573549?v=4" width="100px;"/><br /><sub>Andrea Carraro</sub>](http://www.andreacarraro.it)<br />[💻](https://github.com/toomuchdesign/react-minimal-pie-chart/commits?author=toomuchdesign 'Code') [📖](https://github.com/toomuchdesign/react-minimal-pie-chart/commits?author=toomuchdesign 'Documentation') [🚇](#infra-toomuchdesign 'Infrastructure (Hosting, Build-Tools, etc)') [⚠️](https://github.com/toomuchdesign/react-minimal-pie-chart/commits?author=toomuchdesign 'Tests') [👀](#review-toomuchdesign 'Reviewed Pull Requests') | [<img src="https://avatars3.githubusercontent.com/u/1128559?v=4" width="100px;"/><br /><sub>Stephane Rufer</sub>](https://github.com/rufman)<br />[🐛](https://github.com/toomuchdesign/react-minimal-pie-chart/issues?q=author%3Arufman 'Bug reports') [💻](https://github.com/toomuchdesign/react-minimal-pie-chart/commits?author=rufman 'Code') | [<img src="https://avatars3.githubusercontent.com/u/1413255?v=4" width="100px;"/><br /><sub>Jørgen Aaberg</sub>](https://github.com/jaaberg)<br />[💻](https://github.com/toomuchdesign/react-minimal-pie-chart/commits?author=jaaberg 'Code') | [<img src="https://avatars3.githubusercontent.com/u/16377119?v=4" width="100px;"/><br /><sub>Tobiah Rex</sub>](http://www.tobiahrex.com)<br />[🐛](https://github.com/toomuchdesign/react-minimal-pie-chart/issues?q=author%3ATobiahRex 'Bug reports') | [<img src="https://avatars2.githubusercontent.com/u/11728228?v=4" width="100px;"/><br /><sub>Edward Xiao</sub>](https://edwardxiao.com)<br />[🐛](https://github.com/toomuchdesign/react-minimal-pie-chart/issues?q=author%3Aedwardfhsiao 'Bug reports') |
| :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |


<!-- ALL-CONTRIBUTORS-LIST:END -->


CHANGELOG.md
# 4.0.0

- Forked from original source

# 3.0.2

- Prevent initial animation when component is unmounted

# 3.0.1

- Update `react`/`react-dom` peer dependency version to accept versions `15` and `16`

# 3.0.0

- Make SVG element `display: block` to remove undesired spacing

# 2.0.0

- Compile with Rollup.js to bundle with a transpiled version of `svg-partial-circle v0.2.0`

# 1.1.0

- Migrate React's prop types to external `prop-types` package
- Swap `react-addons-test-utils` with `react-test-renderer` as `enzyme` required dependencies

# 1.0.0

- Enable negative `lengthAngle` to configure clockwise and counterclockwise charts
- Add `rx` and `ry` props to set custom chart center coordinates
- Add `ratio` prop
- Add `radius` prop
- Re-evaluate segments size when `paddingAngle` is > 0
- Make counterclockwise charts by default
- Make <ReactMinimalPieChart> a pure component
- Add eslint parsing

# 0.0.2

- Incorporate `svg-partial-circle` lib

# 0.0.1

- Fix bad imports

# 0.0.0

- Initial release


rollup.config.js

// https://github.com/rollup/rollup-starter-project
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import babel from 'rollup-plugin-babel'

let pkg = require('./package.json')
let external = []
  // Mark dependencies and peerDependencies as external
  .concat(
    Object.keys(pkg.dependencies), Object.keys(pkg.peerDependencies)
  )
  // ...But bundle svg-partial-circle into the package
  .filter(pkgName => pkgName !== 'svg-partial-circle')

let plugins = [
  resolve(),
  babel(),
  commonjs({
    // https://github.com/reactjs/react-redux/issues/643#issuecomment-285008041
    namedExports: {
      'node_modules/react/react.js': ['PureComponent']
    }
  })
]

export default {
  input: 'src/index.js',
  plugins: plugins,
  external: external,
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      sourcemap: true
    },
    {
      file: pkg.module,
      format: 'es',
      sourcemap: true
    }
  ]
}


src/ReactMinimalPieChartPath.js
import React from 'react'
import PropTypes from 'prop-types'
import partialCircle from 'svg-partial-circle'

const PI = Math.PI
const degreesToRadians = degrees => (degrees * PI) / 180

const makePathCommands = (cx, cy, startAngle, lengthAngle, radius) => {
  let patchedLengthAngle = lengthAngle

  if (patchedLengthAngle >= 360) patchedLengthAngle = 359.999
  if (patchedLengthAngle <= -360) patchedLengthAngle = -359.999

  return partialCircle(
    cx,
    cy, // center X and Y
    radius,
    degreesToRadians(startAngle),
    degreesToRadians(startAngle + patchedLengthAngle)
  )
    .map(command => command.join(' '))
    .join(' ')
}

export default function ReactMinimalPieChartPath({
  cx,
  cy,
  startAngle,
  lengthAngle,
  radius,
  lineWidth,
  reveal,
  ...props
}) {
  const actualRadio = radius - lineWidth / 2
  const pathCommands = makePathCommands(cx, cy, startAngle, lengthAngle, actualRadio)
  let strokeDasharray
  let strokeDashoffset

  // Animate/hide paths with "stroke-dasharray" + "stroke-dashoffset"
  // https://css-tricks.com/svg-line-animation-works/
  if (typeof reveal === 'number') {
    strokeDasharray = ((PI * actualRadio) / 180) * Math.abs(lengthAngle)
    strokeDashoffset = strokeDasharray + (strokeDasharray / 100) * reveal
  }

  return (
    <path
      d={pathCommands}
      strokeWidth={lineWidth}
      strokeDasharray={strokeDasharray}
      strokeDashoffset={strokeDashoffset}
      {...props}
    />
  )
}

ReactMinimalPieChartPath.displayName = 'ReactMinimalPieChartPath'

ReactMinimalPieChartPath.propTypes = {
  cx: PropTypes.number.isRequired,
  cy: PropTypes.number.isRequired,
  startAngle: PropTypes.number,
  lengthAngle: PropTypes.number,
  radius: PropTypes.number,
  lineWidth: PropTypes.number,
  reveal: PropTypes.number
}

ReactMinimalPieChartPath.defaultProps = {
  startAngle: 0,
  lengthAngle: 0,
  lineWidth: 100,
  radius: 100
}


src/ReactMinimalPieChart.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Path from './ReactMinimalPieChartPath'

const VIEWBOX_SIZE = 100
const VIEWBOX_HALF_SIZE = VIEWBOX_SIZE / 2

export default class ReactMinimalPieChart extends Component {
  // Class properties
  state = {
    activeSegment: undefined
  }

  // Constructors
  constructor(props) {
    super(props)

    if (this.props.animate === true) {
      this.hideSegments = true
    }

    // Bindings
    this._handlePathClick.bind(this)
  }

  // Private methods
  _startAnimation() {
    this.hideSegments = false
    this.forceUpdate()
  }

  _sumValues(data) {
    return data.reduce((acc, dataEntry) => acc + dataEntry.value, 0)
  }

  _evaluateViewBoxSize(ratio, baseSize) {
    // Wide ratio
    if (ratio > 1) {
      return `${baseSize} ${baseSize / ratio}`
    }
    // Narrow/squared ratio
    return `${baseSize * ratio} ${baseSize}`
  }

  // @TODO extract padding evaluation
  _evaluateDegreesFromValues(data, totalAngle, totalValue, paddingAngle) {
    const total = totalValue || this._sumValues(data)

    // Remove segments padding from total degrees
    const degreesTakenByPadding = paddingAngle * data.length
    let totalDegrees = Math.abs(totalAngle) - degreesTakenByPadding

    if (totalDegrees > 360) totalDegrees = 360
    if (totalAngle < 0) totalDegrees = -totalDegrees

    // Append "degrees" into each data entry
    return data.map(dataEntry => Object.assign({ degrees: (dataEntry.value / total) * totalDegrees }, dataEntry))
  }

  _makeSegmentTransitionStyle(duration, easing) {
    return {
      transition: `stroke-dashoffset ${duration}ms ${easing}`
    }
  }

  _makeSegments(data, props, hide) {
    // Keep track of how many degrees have already been taken
    let lastSegmentAngle = props.startAngle
    const segmentsPaddingAngle = props.paddingAngle * (props.lengthAngle / Math.abs(props.lengthAngle))
    let reveal

    const style = props.animate
      ? this._makeSegmentTransitionStyle(props.animationDuration, props.animationEasing)
      : undefined

    // Hide/reveal the segment?
    if (hide === true) {
      reveal = 0
    } else if (typeof props.reveal === 'number') {
      reveal = props.reveal
    } else if (hide === false) {
      reveal = 100
    }

    return data.map((dataEntry, index) => {
      const startAngle = lastSegmentAngle
      lastSegmentAngle += dataEntry.degrees + segmentsPaddingAngle

      return (
        <Path
          key={dataEntry.key || index}
          cx={props.cx}
          cy={props.cy}
          startAngle={startAngle}
          lengthAngle={dataEntry.degrees}
          radius={props.radius}
          lineWidth={(props.radius / 100) * props.lineWidth}
          reveal={reveal}
          style={style}
          stroke={this.state.activeSegment === index ? dataEntry.highlight || dataEntry.color : dataEntry.color}
          strokeLinecap={props.rounded ? 'round' : undefined}
          fill="none"
          onClick={() => this._handlePathClick(index)}
        />
      )
    })
  }

  _handlePathClick(index) {
    this.setState({
      activeSegment: index
    })

    this.props.clicked(index)
  }

  // Lifecycle methods
  componentDidMount() {
    if (this.props.animate === true && window.requestAnimationFrame) {
      this.initialAnimationTimerId = setTimeout(() => {
        this.initialAnimationTimerId = null
        this.initialAnimationRAFId = window.requestAnimationFrame(() => {
          this.initialAnimationRAFId = null
          this._startAnimation()
        })
      })
    }
  }

  componentWillUnmount() {
    if (this.initialAnimationTimerId) {
      clearTimeout(this.initialAnimationTimerId)
    }
    if (this.initialAnimationRAFId) {
      window.cancelAnimationFrame(this.initialAnimationRAFId)
    }
  }

  render() {
    if (this.props.data === undefined) {
      return null
    }

    const normalizedData = this._evaluateDegreesFromValues(
      this.props.data,
      this.props.lengthAngle,
      this.props.totalValue,
      this.props.paddingAngle
    )

    return (
      <div className={this.props.className} style={this.props.style}>
        <svg
          viewBox={`0 0 ${this._evaluateViewBoxSize(this.props.ratio, VIEWBOX_SIZE)}`}
          width="100%"
          height="100%"
          style={{ display: 'block' }}
        >
          {this._makeSegments(normalizedData, this.props, this.hideSegments)}
        </svg>
        {this.props.children}
      </div>
    )
  }
}

ReactMinimalPieChart.displayName = 'ReactMinimalPieChart'

ReactMinimalPieChart.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.number.isRequired,
      key: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      color: PropTypes.string,
      highlight: PropTypes.string
    })
  ),
  cx: PropTypes.number,
  cy: PropTypes.number,
  ratio: PropTypes.number,
  totalValue: PropTypes.number,
  style: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
  startAngle: PropTypes.number,
  lengthAngle: PropTypes.number,
  paddingAngle: PropTypes.number,
  lineWidth: PropTypes.number,
  radius: PropTypes.number,
  rounded: PropTypes.bool,
  animate: PropTypes.bool,
  animationDuration: PropTypes.number,
  animationEasing: PropTypes.string,
  reveal: PropTypes.number,
  children: PropTypes.node,
  clicked: PropTypes.func
}

ReactMinimalPieChart.defaultProps = {
  cx: VIEWBOX_HALF_SIZE,
  cy: VIEWBOX_HALF_SIZE,
  ratio: 1,
  startAngle: 0,
  lengthAngle: 360,
  paddingAngle: 0,
  lineWidth: 100,
  radius: VIEWBOX_HALF_SIZE,
  rounded: false,
  animate: false,
  animationDuration: 500,
  animationEasing: 'ease-out'
}
